File: src/filters/find.ts
import path from "node:path";
import { cleanText } from "../formatting/ansi.js";
import { limitLines } from "../formatting/limits.js";
import type { FilterOptions, FilterResult } from "./types.js";

const ERROR_PATTERN = /permission denied|no such file|cannot access|find:/i;

export function filterFindOutput(input: string, options: FilterOptions): FilterResult {
  const lines = cleanText(input).split("\n").filter((line) => line.trim());
  const paths: string[] = [];
  const errors: string[] = [];

  for (const line of lines) {
    if (ERROR_PATTERN.test(line)) {
      errors.push(line);
    } else {
      paths.push(line);
    }
  }

  if (paths.length === 0 && errors.length === 0) {
    return {
      output: "",
      kind: "find",
      omitted: false,
      notes: [],
    };
  }

  const directoryCounts = new Map<string, number>();
  for (const filePath of paths) {
    const directory = path.dirname(filePath).replace(/\\/g, "/") || ".";
    directoryCounts.set(directory, (directoryCounts.get(directory) ?? 0) + 1);
  }

  const topDirectories = [...directoryCounts.entries()]
    .sort((left, right) => right[1] - left[1] || left[0].localeCompare(right[0]))
    .slice(0, 8);

  const output: string[] = [`paths: ${paths.length}`];
  if (topDirectories.length > 0) {
    output.push("top directories:");
    for (const [directory, count] of topDirectories) {
      output.push(`  ${directory} (${count})`);
    }
  }

  if (errors.length > 0) {
    output.push(`errors: ${errors.length}`);
    output.push(...errors.slice(0, Math.max(1, options.perFileLines)).map((line) => `  ${line}`));
    if (errors.length > options.perFileLines) {
      output.push(`  ... ${errors.length - options.perFileLines} error lines omitted ...`);
    }
  }

  output.push("sample paths:");
  output.push(...paths.slice(0, Math.max(1, options.perFileLines)).map((filePath) => `  ${filePath}`));
  if (paths.length > options.perFileLines) {
    output.push(`  ... ${paths.length - options.perFileLines} paths omitted ...`);
  }

  const limited = limitLines(output, options.maxLines);
  const omitted =
    paths.length > options.perFileLines ||
    errors.length > options.perFileLines ||
    limited.omitted > 0;

  return {
    output: limited.lines.join("\n"),
    kind: "find",
    omitted,
    notes: omitted ? ["collapsed path listing"] : [],
  };
}

File: internal/filters/find.go
package filters

import (
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/ivanduplenskikh/shrinker/internal/formatting"
)

var findErrorPattern = regexp.MustCompile(`(?i)permission denied|no such file|cannot access|find:`)

func applyFind(input string, options Options) Result {
	lines := strings.Split(formatting.CleanText(input), "\n")
	paths, errs := []string{}, []string{}
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		if findErrorPattern.MatchString(line) {
			errs = append(errs, line)
		} else {
			paths = append(paths, line)
		}
	}
	if len(paths) == 0 && len(errs) == 0 {
		return Result{Kind: "find"}
	}
	counts := map[string]int{}
	for _, name := range paths {
		directory := filepath.ToSlash(filepath.Dir(name))
		if directory == "." || directory == "" {
			directory = "."
		}
		counts[directory]++
	}
	type pair struct {
		name  string
		count int
	}
	top := make([]pair, 0, len(counts))
	for name, count := range counts {
		top = append(top, pair{name, count})
	}
	sort.Slice(top, func(i, j int) bool {
		return top[i].count > top[j].count || (top[i].count == top[j].count && top[i].name < top[j].name)
	})
	if len(top) > 8 {
		top = top[:8]
	}
	output := []string{"paths: " + strconv.Itoa(len(paths))}
	if len(top) > 0 {
		output = append(output, "top directories:")
		for _, item := range top {
			output = append(output, "  "+item.name+" ("+strconv.Itoa(item.count)+")")
		}
	}
	if len(errs) > 0 {
		output = append(output, "errors: "+strconv.Itoa(len(errs)))
		visible := min(len(errs), max(1, options.PerFileLines))
		for _, line := range errs[:visible] {
			output = append(output, "  "+line)
		}
		if visible < len(errs) {
			output = append(output, "  ... "+strconv.Itoa(len(errs)-visible)+" error lines omitted ...")
		}
	}
	output = append(output, "sample paths:")
	visible := min(len(paths), max(1, options.PerFileLines))
	for _, name := range paths[:visible] {
		output = append(output, "  "+name)
	}
	if visible < len(paths) {
		output = append(output, "  ... "+strconv.Itoa(len(paths)-visible)+" paths omitted ...")
	}
	limited, lineOmitted := formatting.LimitLines(output, options.MaxLines)
	return Result{Output: strings.Join(limited, "\n"), Kind: "find", Omitted: len(paths) > options.PerFileLines || len(errs) > options.PerFileLines || lineOmitted > 0}
}
